pkg/util/ensure_disk_controller.go

```diff
@@ -0,0 +1,551 @@
+// controllers are added to the ConfigSpec as necessary for the disks.
+//
+// Please note the following table for the number of controllers of each type
+// that are supported as well as how many disks (per controller) each supports:
+//
+// SATA
+//   - controllers                                    4
+//   - disks                                         30
+//
+// SCSI
+//   - controllers                                    4
+//   - disks (non-paravirtual)                       16
+//   - disks (paravirtual, hardware version <14)     16
+//   - disks (paravirtual, hardware version >=14)   256
+//
+// NVME
+//   - controllers                                    4
+//   - disks (hardware version <20)                  15
+//   - disks (hardware version >=21)                255
+func EnsureDisksHaveControllers(
+	configSpec *vimtypes.VirtualMachineConfigSpec,
+	existingDevices ...vimtypes.BaseVirtualDevice) error {
+
+	if configSpec == nil {
+		return fmt.Errorf("configSpec is nil")
+	}
+
+	var (
+		disks           []*vimtypes.VirtualDisk
+		newDeviceKey    int32
+		pciController   *vimtypes.VirtualPCIController
+		diskControllers = ensureDiskControllerData{
+			controllerKeys:                map[int32]vimtypes.BaseVirtualDevice{},
+			controllerKeysToAttachedDisks: map[int32]int{},
+		}
+	)
+
+	// Inspect the ConfigSpec
+	for i := range configSpec.DeviceChange {
+		var (
+			bdc vimtypes.BaseVirtualDeviceConfigSpec
+			bvd vimtypes.BaseVirtualDevice
+			dc  *vimtypes.VirtualDeviceConfigSpec
+			d   *vimtypes.VirtualDevice
+		)
+
+		if bdc = configSpec.DeviceChange[i]; bdc == nil {
+			continue
+		}
+
+		if dc = bdc.GetVirtualDeviceConfigSpec(); dc == nil {
+			continue
+		}
+
+		if dc.Operation == vimtypes.VirtualDeviceConfigSpecOperationRemove {
+			// Do not consider devices being removed.
+			continue
+		}
+
+		bvd = dc.Device
+		if bvd == nil {
+			continue
+		}
+
+		if d = bvd.GetVirtualDevice(); d == nil {
+			continue
+		}
+
+		switch tvd := bvd.(type) {
+		case *vimtypes.VirtualPCIController:
+			pciController = tvd
+
+		case
+			// SCSI
+			*vimtypes.ParaVirtualSCSIController,
+			*vimtypes.VirtualBusLogicController,
+			*vimtypes.VirtualLsiLogicController,
+			*vimtypes.VirtualLsiLogicSASController,
+			*vimtypes.VirtualSCSIController,
+
+			// SATA
+			*vimtypes.VirtualSATAController,
+			*vimtypes.VirtualAHCIController,
+
+			// NVME
+			*vimtypes.VirtualNVMEController:
+
+			diskControllers.add(bvd)
+
+		case *vimtypes.VirtualDisk:
+
+			disks = append(disks, tvd)
+
+			if controllerKey := d.ControllerKey; controllerKey != 0 {
+				// If the disk points to a controller key, then increment
+				// the number of devices attached to that controller.
+				//
+				// Please note that at this point it is not yet known if the
+				// controller key is a *valid* controller.
+				diskControllers.attach(controllerKey)
+			}
+		}
+
+		// Keep track of the largest device key used.
+		if d.Key < newDeviceKey {
+			newDeviceKey = d.Key
+		}
```

Comment:
Looks like the code is doing the opposite of what the comment says?

Reply:
Largest in the sense of largest *negative* number. I could say smallest to be technically accurate, but I thought this was more clear. Suggestions on how to phrase this?

Reply:
Ah, I didn't realize that device keys in a ConfigSpec are negative numbers, so I was confused.
What you have updated looks good to me.